File: plugin/Functions/CurveOperations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using Rhino;
using Rhino.DocObjects;
using Rhino.Geometry;
using Rhino.Geometry.Intersect;

namespace RhinoMCPPlugin.Functions;

/// <summary>
/// Curve operations: Project, Intersect, Split
/// </summary>
public partial class RhinoMCPFunctions
{
    /// <summary>
    /// Project a curve onto surfaces or polysurfaces.
    /// </summary>
    [McpCommand("project_curve")]
    public JObject ProjectCurve(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveId = parameters["curve_id"]?.ToString();
        var targetIds = parameters["target_ids"]?.ToObject<List<string>>();
        var direction = parameters["direction"]?.ToObject<List<double>>();
        var name = parameters["name"]?.ToString();

        if (string.IsNullOrEmpty(curveId))
            throw new ArgumentException("Project requires a curve_id");

        if (targetIds == null || targetIds.Count == 0)
            throw new ArgumentException("Project requires at least one target surface/polysurface ID");

        if (direction == null || direction.Count != 3)
            throw new ArgumentException("Project requires a direction vector [x, y, z]");

        var curveObj = doc.Objects.Find(new Guid(curveId));
        if (curveObj == null)
            throw new InvalidOperationException($"Curve object with ID {curveId} not found");

        var curve = curveObj.Geometry as Curve;
        if (curve == null)
            throw new InvalidOperationException($"Object {curveId} is not a curve");

        var directionVec = new Vector3d(direction[0], direction[1], direction[2]);
        var breps = new List<Brep>();
        var meshes = new List<Mesh>();

        foreach (var idStr in targetIds)
        {
            var obj = doc.Objects.Find(new Guid(idStr));
            if (obj == null) continue;

            if (obj.Geometry is Brep brep)
                breps.Add(brep);
            else if (obj.Geometry is Mesh mesh)
                meshes.Add(mesh);
            else if (obj.Geometry is Extrusion extrusion)
                breps.Add(extrusion.ToBrep());
        }

        if (breps.Count == 0 && meshes.Count == 0)
            throw new InvalidOperationException("No valid target surfaces or meshes found");

        var projectedCurves = new List<Curve>();
        
        if (breps.Count > 0)
        {
            var result = Curve.ProjectToBrep(curve, breps, directionVec, doc.ModelAbsoluteTolerance);
            if (result != null) projectedCurves.AddRange(result);
        }

        if (meshes.Count > 0)
        {
            var result = Curve.ProjectToMesh(curve, meshes, directionVec, doc.ModelAbsoluteTolerance);
            if (result != null) projectedCurves.AddRange(result);
        }

        if (projectedCurves.Count == 0)
            return new JObject
            {
                ["result_ids"] = new JArray(),
                ["message"] = "Projection failed - no intersections found"
            };

        var resultIds = new JArray();
        foreach (var projCurve in projectedCurves)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddCurve(projCurve, attr);
            resultIds.Add(id.ToString());
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = projectedCurves.Count,
            ["message"] = $"Projected {projectedCurves.Count} curve(s)"
        };
    }

    /// <summary>
    /// Find intersection points or curves between two curves.
    /// </summary>
    [McpCommand("intersect_curves")]
    public JObject IntersectCurves(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveIdA = parameters["curve_id_a"]?.ToString();
        var curveIdB = parameters["curve_id_b"]?.ToString();
        var tolerance = parameters["tolerance"]?.ToObject<double>() ?? doc.ModelAbsoluteTolerance;
        var name = parameters["name"]?.ToString();

        if (string.IsNullOrEmpty(curveIdA) || string.IsNullOrEmpty(curveIdB))
            throw new ArgumentException("Intersect requires two curve IDs (curve_id_a and curve_id_b)");

        var objA = doc.Objects.Find(new Guid(curveIdA));
        var objB = doc.Objects.Find(new Guid(curveIdB));

        if (objA == null || objB == null)
            throw new InvalidOperationException("One or both curve objects not found");

        var curveA = objA.Geometry as Curve;
        var curveB = objB.Geometry as Curve;

        if (curveA == null || curveB == null)
            throw new InvalidOperationException("One or both objects are not curves");

        var intersections = Intersection.CurveCurve(curveA, curveB, tolerance, tolerance);

        if (intersections == null || intersections.Count == 0)
            return new JObject
            {
                ["points"] = new JArray(),
                ["point_ids"] = new JArray(),
                ["curve_ids"] = new JArray(),
                ["message"] = "No intersections found"
            };

        var resultPointIds = new JArray();
        var resultCurveIds = new JArray();
        var resultPoints = new JArray();

        foreach (var eventItem in intersections)
        {
            if (eventItem.IsPoint)
            {
                var pt = eventItem.PointA;
                resultPoints.Add(new JArray { pt.X, pt.Y, pt.Z });

                var attr = new ObjectAttributes();
                if (!string.IsNullOrEmpty(name))
                    attr.Name = name + "_point";

                var id = doc.Objects.AddPoint(pt, attr);
                resultPointIds.Add(id.ToString());
            }
            else if (eventItem.IsOverlap)
            {
                // For overlaps, we can extract the overlapping segment from curve A
                var overlapCurve = curveA.Trim(eventItem.OverlapA.T0, eventItem.OverlapA.T1);
                if (overlapCurve != null)
                {
                    var attr = new ObjectAttributes();
                    if (!string.IsNullOrEmpty(name))
                        attr.Name = name + "_overlap";

                    var id = doc.Objects.AddCurve(overlapCurve, attr);
                    resultCurveIds.Add(id.ToString());
                }
            }
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["points"] = resultPoints,
            ["point_ids"] = resultPointIds,
            ["curve_ids"] = resultCurveIds,
            ["count"] = intersections.Count,
            ["message"] = $"Found {intersections.Count} intersection(s)"
        };
    }

    /// <summary>
    /// Split a curve at specified parameters or points.
    /// </summary>
    [McpCommand("split_curve")]
    public JObject SplitCurve(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        var curveId = parameters["curve_id"]?.ToString();
        var splitParams = parameters["parameters"]?.ToObject<List<double>>();
        var pointIds = parameters["point_ids"]?.ToObject<List<string>>();
        var deleteSource = parameters["delete_source"]?.ToObject<bool>() ?? true;
        var name = parameters["name"]?.ToString();

        if (string.IsNullOrEmpty(curveId))
            throw new ArgumentException("Split requires a curve_id");

        var curveObj = doc.Objects.Find(new Guid(curveId));
        if (curveObj == null)
            throw new InvalidOperationException($"Curve object with ID {curveId} not found");

        var curve = curveObj.Geometry as Curve;
        if (curve == null)
            throw new InvalidOperationException($"Object {curveId} is not a curve");

        var parametersToSplit = new List<double>();

        if (splitParams != null && splitParams.Count > 0)
        {
            parametersToSplit.AddRange(splitParams);
        }

        if (pointIds != null && pointIds.Count > 0)
        {
            foreach (var idStr in pointIds)
            {
                var obj = doc.Objects.Find(new Guid(idStr));
                if (obj == null) continue;

                Point3d pt;
                if (obj.Geometry is Point point)
                    pt = point.Location;
                else if (obj.Geometry is TextDot dot) // Dot is different, but usually we deal with Point
                    pt = dot.Point;
                else
                    continue;

                double t;
                if (curve.ClosestPoint(pt, out t, doc.ModelAbsoluteTolerance))
                {
                    parametersToSplit.Add(t);
                }
            }
        }

        if (parametersToSplit.Count == 0)
            throw new ArgumentException("No split parameters or points provided");

        var splitCurves = curve.Split(parametersToSplit);

        if (splitCurves == null || splitCurves.Length == 0)
            throw new InvalidOperationException("Split operation failed");

        var resultIds = new JArray();
        foreach (var segment in splitCurves)
        {
            var attr = new ObjectAttributes();
            if (!string.IsNullOrEmpty(name))
                attr.Name = name;

            var id = doc.Objects.AddCurve(segment, attr);
            resultIds.Add(id.ToString());
        }

        if (deleteSource)
        {
            doc.Objects.Delete(curveObj, true);
        }

        doc.Views.Redraw();

        return new JObject
        {
            ["result_ids"] = resultIds,
            ["count"] = splitCurves.Length,
            ["message"] = $"Split curve into {splitCurves.Length} segment(s)"
        };
    }
}

File: server/src/rhinomcp/tools/curve_operations.py
"""
Curve operations for advanced geometry manipulation.

These tools provide interfaces for:
- Project Curve: Project a curve onto surfaces or meshes
- Intersect Curves: Find intersections between two curves
- Split Curve: Split a curve at parameters or points
"""

from mcp.server.fastmcp import Context
from rhinomcp.server import get_rhino_connection, mcp, logger
from typing import List, Optional, Dict, Any


@mcp.tool()
def project_curve(
    ctx: Context,
    curve_id: str,
    target_ids: List[str],
    direction: List[float],
    name: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Project a curve onto surfaces, polysurfaces, or meshes.

    Parameters:
    - curve_id: The curve ID (GUID) to project
    - target_ids: List of surface, polysurface, or mesh IDs to project onto
    - direction: Projection direction as [x, y, z] vector
    - name: Optional name for the resulting curves

    Returns:
    - Dictionary with result_ids (list of created curve IDs) and message
    """
    try:
        if len(direction) != 3:
            return {"success": False, "message": "Direction must be [x, y, z] vector"}
        if not target_ids:
            return {"success": False, "message": "At least one target ID is required"}

        rhino = get_rhino_connection()
        params = {
            "curve_id": curve_id,
            "target_ids": target_ids,
            "direction": direction,
        }
        if name:
            params["name"] = name

        result = rhino.send_command("project_curve", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Curve projected successfully")
        }
    except Exception as e:
        logger.error(f"Error in project_curve: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def intersect_curves(
    ctx: Context,
    curve_id_a: str,
    curve_id_b: str,
    tolerance: Optional[float] = None,
    name: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Find intersection points or curves between two curves.

    Parameters:
    - curve_id_a: The first curve ID
    - curve_id_b: The second curve ID
    - tolerance: Optional intersection tolerance (defaults to document tolerance)
    - name: Optional base name for created intersection objects

    Returns:
    - Dictionary with point_ids, curve_ids, and message
    """
    try:
        rhino = get_rhino_connection()
        params = {
            "curve_id_a": curve_id_a,
            "curve_id_b": curve_id_b,
        }
        if tolerance is not None:
            params["tolerance"] = tolerance
        if name:
            params["name"] = name

        result = rhino.send_command("intersect_curves", params)
        return {
            "success": True,
            "point_ids": result.get("point_ids", []),
            "curve_ids": result.get("curve_ids", []),
            "points": result.get("points", []),
            "message": result.get("message", "Intersections found")
        }
    except Exception as e:
        logger.error(f"Error in intersect_curves: {str(e)}")
        return {"success": False, "message": str(e)}


@mcp.tool()
def split_curve(
    ctx: Context,
    curve_id: str,
    parameters: Optional[List[float]] = None,
    point_ids: Optional[List[str]] = None,
    delete_source: bool = True,
    name: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Split a curve at specified parameters or points.

    Parameters:
    - curve_id: The curve ID to split
    - parameters: Optional list of curve parameters to split at
    - point_ids: Optional list of point object IDs to split at (closest points on curve will be used)
    - delete_source: Whether to delete the original curve (default: True)
    - name: Optional name for the resulting curve segments

    Returns:
    - Dictionary with result_ids (list of created curve segment IDs) and message
    """
    try:
        if not parameters and not point_ids:
            return {"success": False, "message": "Either parameters or point_ids must be provided"}

        rhino = get_rhino_connection()
        params = {
            "curve_id": curve_id,
            "delete_source": delete_source,
        }
        if parameters:
            params["parameters"] = parameters
        if point_ids:
            params["point_ids"] = point_ids
        if name:
            params["name"] = name

        result = rhino.send_command("split_curve", params)
        return {
            "success": True,
            "result_ids": result.get("result_ids", []),
            "message": result.get("message", "Curve split successfully")
        }
    except Exception as e:
        logger.error(f"Error in split_curve: {str(e)}")
        return {"success": False, "message": str(e)}
